crates/accelerate/src/inverse_cancellation.rs

```diff
@@ -0,0 +1,193 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use ahash::RandomState;
+use hashbrown::HashSet;
+use indexmap::IndexMap;
+use pyo3::prelude::*;
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+
+use qiskit_circuit::circuit_instruction::OperationFromPython;
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType};
+use qiskit_circuit::operations::Operation;
+use qiskit_circuit::packed_instruction::PackedInstruction;
+
+fn gate_eq(py: Python, gate_a: &PackedInstruction, gate_b: &OperationFromPython) -> PyResult<bool> {
+    if gate_a.op.name() != gate_b.operation.name() {
+        return Ok(false);
+    }
+    let a_params = gate_a.params_view();
+    if a_params.len() != gate_b.params.len() {
+        return Ok(false);
+    }
+    let mut param_eq = true;
+    for (a, b) in a_params.iter().zip(&gate_b.params) {
+        if !a.is_close(py, b, 1e-10)? {
+            param_eq = false;
+            break;
+        }
+    }
+    Ok(param_eq)
+}
+
+fn run_on_self_inverse(
+    py: Python,
+    dag: &mut DAGCircuit,
+    op_counts: &IndexMap<String, usize, RandomState>,
+    self_inverse_gate_names: HashSet<String>,
+    self_inverse_gates: Vec<OperationFromPython>,
+) -> PyResult<()> {
+    if !self_inverse_gate_names
+        .iter()
+        .any(|name| op_counts.contains_key(name))
+    {
+        return Ok(());
+    }
+    for gate in self_inverse_gates {
+        let gate_count = op_counts.get(gate.operation.name()).unwrap_or(&0);
+        if *gate_count <= 1 {
+            continue;
+        }
+        let mut collect_set: HashSet<String> = HashSet::with_capacity(1);
+        collect_set.insert(gate.operation.name().to_string());
+        let gate_runs: Vec<Vec<NodeIndex>> = dag.collect_runs(collect_set).unwrap().collect();
```

Comment:
It looks like `DAGCircuit::collect_runs` returns an `Option` only because it directly returns the result of the Rustworkx core function, which uses `None` to represent a cycle in the graph (...which is questionable lol). 

Perhaps it is beyond the scope of this PR, but it might be nice if you can change `DAGCircuit::collect_runs` to unwrap that option internally. If it fails, it should be on `DAGCircuit` for getting itself into an invalid state.

Reply:
Yeah, I'm fine with making this change to `collect_runs()` but I'd say lets save this for a follow-up PR so we can do it in isolation.